### PR TITLE
Adds support for various returned objects in define method

### PR DESCRIPTION
**Pull request by eahefnawy**
In order to make it easier to programmatically define components we should make this function more like defining components in serverless.yml.
```
define() {
  return {
    foo: {
      type: ‘FooComponent’,
      inputs: ...
    },
    bar: {
      type: ‘./path/to/BarComponent’,
      inputs: ...
    }
  }
}
```
We should also support pointing to types that are already loaded. 

const FooComponent = await context.loadType(‘FooComponent’)
```
define() {
  return {
    foo: {
      type: FooComponent,
      inputs: ...
    }
  }
}
```
As well as component instances that have already been constructed

const FooComponent = await context.loadType(‘FooComponent’)
```
define() {
  return {
    foo: context.construct(FooComponent, {})
  }
}
```

### CODE REVIEW

File: registry/AwsProvider/src/index.js [hunk 1]
```diff
@@ -27,10 +26,6 @@ const AwsProvider = (SuperClass) =>
       if (!/.+-.+.\d+/.test(this.region)) {
         throw new Error(`Invalid region "${this.region}" in your AWS provider setup`)
       }
-
-      if (!this.credentials || isEmpty(this.credentials)) {
```
Comment (pmuens) [inline registry/AwsProvider/src/index.js]: This should be kept and extended to support environment based config rather than removed.

Furthermore it should be part of a different PR.

Comment (brianneisler) [reply registry/AwsProvider/src/index.js]: Agreed this should be part of a different PR

File: src/utils/component/defineComponent.js [hunk 2]
```diff
@@ -18,6 +16,29 @@ const defineComponent = async (component, state, context) => {
...
+    //     context,
+    //     root: context.root,
+    //     path: context.root
+    //   },
+    //   context
+    // )
+    let typeQueryChildren = filter(isTypeConstruct, map(resolve, children))
+
+    typeQueryChildren = await map(async (typeQueryChild) => {
+      const child = await context.loadType(typeQueryChild.type)
```
Comment (pmuens) [inline src/utils/component/defineComponent.js]: Just stumbled upon this one today... This needs to be renamed to `context.import`

**Comment by pmuens** [comment]
This PR is related to #362

**Comment by brianneisler** [comment]
@eahefnawy i believe @pmuens PR has covered what this PR originally intended to provide. We can probably close this PR in favor of https://github.com/serverless/components/pull/362

@pmuens one note is that we should also copy over the change to the `loadType` method from this PR that simply returns the type if the query is already a type.

**Comment by eahefnawy** [comment]
closing, but not deleting branch. I'll leave it for @pmuens to copy any needed code from here.

**Comment by pmuens** [comment]
FYI: @brianneisler @eahefnawy just ported the changes which were made to `loadType` over into #362 and added a unit tests as well.
